[Async object upload 3] Invoke ChunkUploader from bucket implementations

gargnitingoogle opened the pull request:
### Description
This adds implementation of gcs.ChunkUploader interface and CreateChunkUploader for all bucket types, and their corresponding unit tests.

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - All unit tests passed.
3. Integration tests - Ran as presubmit

sethiay commented on internal/storage/debug_bucket.go:
```diff
@@ -163,8 +163,12 @@ func (b *debugBucket) CreateChunkUploader(
-	progressFunc func(int64)) (gcs.ChunkUploader, error) {
-	return nil, fmt.Errorf("not implemented yet")
+	progressFunc func(int64)) (uploader gcs.ChunkUploader, err error) {
+	id, desc, start := b.startRequest("CreateChunkUploader(%q)", req.Name)
```
please check if we should do similar to NewReader.

gargnitingoogle replied:
Good point. Added debugChunkUploader similar to debugReader.

sethiay reviewed (commented):
I will review the fake bucket.go and mock_bucket.go.

gargnitingoogle commented on internal/monitor/bucket.go:
```diff
@@ -142,7 +142,10 @@ func (mb *monitoringBucket) CreateChunkUploader(
 	progressFunc func(int64)) (gcs.ChunkUploader, error) {
-	return nil, fmt.Errorf("not implemented yet")
+	startTime := time.Now()
+	o, err := mb.wrapped.CreateChunkUploader(ctx, req, writeChunkSize, progressFunc)
```
I tried to do this, but realized that the scope for that will be a lot bigger than just here. We currently only monitor for bytesBeingRead and have registered the same at opencensus. For chunk-uploader, we'll have to add bytesBeingWritten and corresponding changes on the viewer. I'll add monitoring for that later. Currently adding a TODO for that here.

gargnitingoogle replied:
Captured in internal bug #308572499.

gargnitingoogle reviewed (commented):
Addressed ayush' comments

gargnitingoogle commented:
@sethiay addressed all your comments. There is one request that I can't do here, namely the monitoring for chunkuploader.Upload() calls as the changes for that are much bigger than what can be addressed in this single PR. For now, have added a TODO for it. PTAL

gargnitingoogle commented:
> I will review the fake bucket.go and mock_bucket.go.

@sethiay what does this mean ? Have you reviewed only these two files, or haven't reviewed these two files yet?

sethiay commented on internal/storage/bucket_handle_test.go:
```diff
@@ -366,6 +367,56 @@ func (t *BucketHandleTest) TestCreateObjectMethodWithGenerationAsZeroWhenObjectA
 	AssertTrue(errors.As(err, &precondition))
 }
 
+func (t *BucketHandleTest) TestCreateChunkUploaderMethodWithValidObject() {
```
can we also write a test case or append to this test case where we also call upload and close form chunk uploader created here?

gargnitingoogle replied:
Added Upload, Close and corresponding asserts to both the positive tests.

sethiay commented on internal/storage/bucket_handle_test.go:
```diff
@@ -366,6 +367,56 @@ func (t *BucketHandleTest) TestCreateObjectMethodWithGenerationAsZeroWhenObjectA
+func (t *BucketHandleTest) TestCreateChunkUploaderMethodWithGenerationAsZero() {
+	var generation int64 = 0
+	chunkSize := googleapi.MinUploadChunkSize
+	uploader, err := t.bucketHandle.CreateChunkUploader(context.Background(),
```
can we follow arrange, act and assert here and everywhere go/unit-testing-practices?polyglot=go#structure?

gargnitingoogle replied:
I only followed the pattern of the other tests in this file. 
I have updated the 3 chunkUploader tests that I've added. I have followed the arrange,act,assert pattern everywhere in any new files.

sethiay commented on internal/storage/debug_bucket.go:
```diff
@@ -159,12 +159,67 @@ func (b *debugBucket) CreateObject(
+func (dcu *debugChunkUploader) Upload(ctx context.Context, contents io.Reader) error {
+	err := dcu.wrapped.Upload(ctx, contents)
+	// Don't log EOF errors, which are par for the course.
+	if err != nil && err != io.EOF {
```
we will not get io.IOF errors from upload, right ? if yes, can we remove this check ?

gargnitingoogle replied:
Yes, removed.

sethiay commented on internal/storage/debug_bucket.go:
```diff
@@ -159,12 +159,67 @@ func (b *debugBucket) CreateObject(
+
+	uploader, err = b.wrapped.CreateChunkUploader(ctx, req, writeChunkSize, progressFunc)
+	if err != nil {
+		uploader = nil
```
remove lint errors.

gargnitingoogle replied:
Done

sethiay commented on internal/storage/caching/fast_stat_bucket_test.go:
```diff
@@ -135,6 +135,80 @@ func (t *CreateObjectTest) WrappedSucceeds() {
+
+	// Wrapped
+	ExpectCall(t.wrapped, "CreateChunkUploader")(Any(), Any(), Any(), Any()).
+		WillOnce(Return(nil, errors.New("taco")))
```
nit: put "taco" in a constant inside method similar to CallsEraseAndWrapped

gargnitingoogle replied:
Done.

sethiay commented on internal/storage/caching/fast_stat_bucket_test.go:
```diff
@@ -135,6 +135,80 @@ func (t *CreateObjectTest) WrappedSucceeds() {
+	ExpectCall(t.cache, "Insert")(obj, timeutil.TimeEq(t.clock.Now().Add(ttl)))
+
+	// Call
+	o, err := t.bucket.CreateObject(context.TODO(), &gcs.CreateObjectRequest{})
```
change CreateObject to CreateChunkUploader as we are not testing CreateObject here.

gargnitingoogle replied:
I tried doing this, but this test needs the return values to be mocked, which is not possible for chunkUploader as it's an interface. In other words, a precreated instance of ChunkUploader needs to be passed here which can be updated within this call. 
From what I see, mocking is only possible for structs and primitives, but not for interface return-type values. This is probably why there are no tests for NewReader here as it returns io.Reader which is also an interface.
So, this wrapSucceeds test can't be implemented, unless we can find a way to mock ChunkUploader interface return-type.

gargnitingoogle replied:
Disabling this test for now. Will remove the code once we have an agreement.

gargnitingoogle replied:
Removed the test.

sethiay commented:
> > I will review the fake bucket.go and mock_bucket.go.
> 
> @sethiay what does this mean ? Have you reviewed only these two files, or haven't reviewed these two files yet?

It meant the latter. I have reviewed these two files also now.

gargnitingoogle commented on internal/gcsx/content_type_bucket_test.go:
```diff
@@ -104,6 +105,40 @@ func TestContentTypeBucket_CreateObject(t *testing.T) {
+		}
+
+		o, err := uploader.Close(context.Background())
+		if o == nil || nil != err {
```
DOne

gargnitingoogle commented:
> > > I will review the fake bucket.go and mock_bucket.go.
> > 
> > 
> > @sethiay what does this mean ? Have you reviewed only these two files, or haven't reviewed these two files yet?
> 
> It meant the latter. I have reviewed these two files also now.

ok. Thanks.

@sethiay I've addressed all your comments. All but  2 are closed. PTAL.

gargnitingoogle commented:
@sethiay @ashmeenkaur please let me know if there is any other comment. If not, please approve.

ashmeenkaur commented on internal/storage/caching/fast_stat_bucket_test.go:
```diff
@@ -135,6 +135,92 @@ func (t *CreateObjectTest) WrappedSucceeds() {
+	ExpectThat(err, Error(HasSubstr(name)))
+}
+
+// func (t *CreateChunkUploaderTest) WrappedSucceeds() {
```
Is this supposed to be commented?

gargnitingoogle replied:
Why I can't implement this particular test case is in https://github.com/GoogleCloudPlatform/gcsfuse/pull/1417#discussion_r1377425289 . So I had to disable it.

gargnitingoogle replied:
Removed the commented out test.

gargnitingoogle commented:
@sethiay @ashmeenkaur please review.

sethiay commented on internal/storage/bucket_handle.go:
```diff
@@ -156,6 +154,13 @@ func (bh *bucketHandle) CreateObject(ctx context.Context, req *gcs.CreateObjectR
 		obj = obj.If(preconditions)
 	}
 
+	return obj, (req.GenerationPrecondition == nil || preconditions.DoesNotExist)
```
can we please add a comment ?

gargnitingoogle replied:
Good idea. Will do.

gargnitingoogle replied:
Added comment explaining what the bool is for and how it's calculated.

sethiay commented on internal/storage/bucket_handle_test.go:
```diff
@@ -366,6 +367,89 @@ func (t *BucketHandleTest) TestCreateObjectMethodWithGenerationAsZeroWhenObjectA
+
+	AssertEq(nil, uploader)
+	AssertNe(nil, err)
+	if !strings.Contains(err.Error(), expectedErrorSubstring) {
```
nit: AssertTrue(strings.Contains(err.Error(), expectedErrorSubstring))

gargnitingoogle replied:
This is on purpose. This is to make it easier to identify the issue directly from the error log, by printing out the expected error log and the actual error log, rather than having to debug or change the code to find out the issue. I'll keep it this way.

gargnitingoogle reviewed (commented):
Resolved open comment.

gargnitingoogle commented:
@sethiay I resolved your comments. Please approve.
